p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,351 @@
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]node.Node, error) {
+	const extendedSquareSize = 256
+	br := bufio.NewReader(r)
+	nodes := make([]node.Node, 0, extendedSquareSize)
+	nodeCollector := func(hash []byte, children ...[]byte) {
+		cid, err := cidFromNamespacedSha256(hash)
+		if err != nil {
+			panic(fmt.Sprintf("nmt lib created a malformed hash: %s", err))
+		}
+		switch len(children) {
+		case 1:
+			prependNode(nmtLeafNode{
+				cid:  cid,
+				Data: children[0],
+			}, &nodes)
+		case 2:
+			prependNode(nmtNode{
+				cid: cid,
+				l:   children[0],
+				r:   children[1],
+			}, &nodes)
+		default:
+			panic("expected a binary tree")
+		}
+	}
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(nodeCollector),
+	)
+	for {
+		share := make([]byte, shareSize+namespaceSize)
+		if _, err := io.ReadFull(br, share); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if err := n.Push(share[:namespaceSize], share[namespaceSize:]); err != nil {
+			return nil, err
+		}
+	}
+	// to trigger the collection of nodes:
+	_ = n.Root()
+	return nodes, nil
+}
+
+func prependNode(newNode node.Node, nodes *[]node.Node) {
+	*nodes = append(*nodes, node.Node(nil))
+	copy((*nodes)[1:], *nodes)
+	(*nodes)[0] = newNode
+}
+
+func NmtNodeParser(block blocks.Block) (node.Node, error) {
+	const (
+		// length of the domain separator for leaf and inner nodes:
+		prefixOffset = 1
+		// nmtHashSize = flagSize+sha256.Size
+		nmtHashSize = namespaceSize*2 + sha256.Size
+	)
+	var (
+		leafPrefix  = []byte{nmt.LeafPrefix}
+		innerPrefix = []byte{nmt.NodePrefix}
+	)
+	data := block.RawData()
+	if len(data) == 0 {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: nil,
+		}, nil
+	}
+	domainSeparator := data[:prefixOffset]
+	if bytes.Equal(domainSeparator, leafPrefix) {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: data[prefixOffset:],
+		}, nil
+	}
+	if bytes.Equal(domainSeparator, innerPrefix) {
+		return nmtNode{
+			cid: block.Cid(),
+			l:   data[prefixOffset : prefixOffset+nmtHashSize],
+			r:   data[prefixOffset+nmtHashSize:],
+		}, nil
+	}
+	return nil, fmt.Errorf(
+		"expected first byte of block to be either the leaf or inner node prefix: (%x, %x), got: %x)",
+		leafPrefix,
+		innerPrefix,
+		domainSeparator,
+	)
+}
+
+var _ node.Node = (*nmtNode)(nil)
+var _ node.Node = (*nmtLeafNode)(nil)
+
+type nmtNode struct {
+	// TODO(ismail): we might want to export these later
+	cid  cid.Cid
+	l, r []byte
+}
+
+func (n nmtNode) RawData() []byte {
+	return append([]byte{nmt.NodePrefix}, append(n.l, n.r...)...)
+}
+
+func (n nmtNode) Cid() cid.Cid {
+	return n.cid
+}
+
+func (n nmtNode) String() string {
+	return fmt.Sprintf(`
+node {
+	hash: %x,
+	l: %x,
+	r: %x"
+}`, n.cid.Hash(), n.l, n.r)
+}
+
+func (n nmtNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (n nmtNode) Resolve(path []string) (interface{}, []string, error) {
```

Comment:
I've implemented it this way around because of the comment on the interface method:
```go
// ResolveLink is a helper function that calls resolve and asserts the
// output is a link
ResolveLink(path []string) (*Link, []string, error)
```

Also, it seems like other plugins do it the same way around (e.g. go-ipld-git and others).

Reply:
Take a look at the `ProtoNode` that implements my suggestion. Also, take it for reference, as it the most mature IPLD node.

Reply:
Resolving suggestion as we do not want to depend on protobuf inside of the NMT.